docs/en/sql-reference/functions/string-search-functions.md

```diff
@@ -901,3 +901,42 @@ SELECT
 │ 100                                          │ 200                                          │ 100-200                                      │ 100                                       │
 └──────────────────────────────────────────────┴──────────────────────────────────────────────┴──────────────────────────────────────────────┴───────────────────────────────────────────┘
 ```
+
+
+## regexpExtractAll(haystack, pattern[, index])
```

Comment:
Question about functions `regexpExtractAll` and `regexpExtract` (which was recently added): ClickHouse already has functions `extract` and `extractAll` which are also based on re2 regexps. Why didn't we extend them with an optional `index` parameter instead of adding two new functions and associated code?

Reply:
Extending already existed functions extract/exactAll will break their backward compatiability. 


The default behavior of spark regexp_extract is different from ch extract. In spark `regexp_extract(haystack, pattern[, index])`, the third argument is optional, and its default value is 1, which means `regexp_extract(haystack, pattern)` always returns the first captured matched string in haystack. But ch `extract(haystack, pattern)` returns the segment that matches the entire pattern if their is no capture in pattern, otherwise returns the first captured matched string.

Reply:
For example 
In spark: `select regexp_extract('100-200', '\\d+-\\d+')` throws exception `Regex group count is 0, but the specified group index is 1`. 
But in CH: `select extract('100-200', '\\d+-\\d+')` returns `100-200`. 
This issue can't be solved by simly extending existing function extract, so is extractAll.

Reply:
Thanks for the explanation!

I think there is a trade-off to be made. On the one hand, Spark compatibility in ClickHouse is great-to-have. Even though I think that at no point one will be able to take Spark SQL and execute it unchanged in ClickHouse with the same result. In such cases, migration/compatibility hints in the docs will already be helpful. Note that Spark is just one out of many other databases, and supporting the SQL dialects of all of them is out of scope. My preference would be to check SQL dialects in other databases for capabilities missing in ClickHouse (e.g. in terms of convenience or expressiveness) and then to provide those in ClickHouse but rather as an extension of the existing ClickHouse functions. My concern with replicating "foreign" features with clean-slate implementations like for `regexp_replace(_all)` is that we can end up with functions which provide a superset of other functions and the amount of necessary implementation.

Long story short, IMHO the more sustainable approach is this:
1. Extend the existing function `extract` with an optional parameter `index`. The parameter can take these values:
- `index = -1` (the default): the current behavior, i.e. if the pattern has no capturing group then match against the pattern, if the pattern has capturing group(s) then match against the 1st capturing group
- `index = 0` (corresponds to 0 in Spark): always match against the whole pattern irrespective of subgroups
- `index = 1/2/...` (corresponds to 1/2/... in Spark): match against capturing group 1/2/... (*)
2. Explain in ClickHouse docs that `index =0/1/2/...` corresponds to Spark's `regexp_extract`.
3. Possibly introduce a SQL function `regex_extract` but it should really only be a frontend/wrapper around `extract`.

My preference would be to handle `regexp_extract_all`/`extractAll` exactly the same way.

As a result, we win three times: 1. none of the existing use cases will break. 2. Users transferring from Spark will have an equally easy time. 3. Much less code needs to be added and maintained.

I know that you already invested time and effort to prepare this PR, and redoing it is a pitty of course. On the other hand, having a clean long-term solution is always better for users and developers.

(*) As a side question: what happens if `index = N > 0` and the pattern contains less than `N` capturing groups?? Guess we could throw in this case.